### PR TITLE
Code refactoring

### LINKED ISSUE #5: Consider moving training related parameters to `fit()`
Things like weight decay, batch size, and epochs may better belong in the `fit` method.
https://github.com/DAI-Lab/CTGAN/blob/94ac76b3ff4dae827eb626abe65923150633e359/ctgan/ctgan_model.py#L426-L432

### LINKED ISSUE #10: Reorganize the project structure
A few changes can be made to improve the overall code maintainability:

* Modularize the code, separating code dedicated to specific topics on different modules.
* Split some parts of the code in smaller, single purpose methods.
* Rename the `Cond` class to `ConditionalGenerator`.
* Make `random_choice_prob_index` a method of the `ConditionalGenerator` class.
* Make `calc_gradient_penalty` a method of the `Discriminator` class.
* Make `cond_loss` a `CTGANSynthesizer` method.
* Make a `_build_models` method where all the internal models are built. See #4
* Separate the fitting process in two methods:
    * One with categorical variables, which uses the `cond` vector
    * Another one without categorical variables.
* Rename some variables to be a bit more verbose
* Replace plain asserts with user more verbose and type specific exceptions
* Add docstrings to all the public methods
* Remake `cli.py` as `__main__.py` and use `argparse` instead of `absl`
* Move `epochs` arg to `fit` method. See #5

The new project structure should look as follows:

```
ctgan
├── __init__.py
├── __main__.py: New CLI module
├── conditional.py: ConditionalGenerator
├── data.py: Methods to write and read data
├── models.py: Discriminator, Generator and Residual classes
├── sampler.py: Sampler class
├── synthesizer.py: CTGANSynthesizer class
└── transformer.py: DataTransformer class
```

### PR DESCRIPTION
**Pull request by csala**
Resolve #10 : Reorganize the project structure and do some usability improvements.
Also resolve #5 .

The complete list of changes made is:

* Modularize the code, separating code dedicated to specific topics on different modules.
* Rename the `Cond` class to `ConditionalGenerator`.
* Make `random_choice_prob_index` a method of the `ConditionalGenerator` class.
* Make `calc_gradient_penalty` a method of the `Discriminator` class.
* Make `cond_loss` a `CTGANSynthesizer` method.
* Reimplement some parts of the  `DataTransformer` to use a OneHotEncoder and also support categorical variables through the python interface.
* Remake `cli.py` as `__main__.py` and use `argparse` instead of `absl`
* Add support for CSV data and JSON Metadata. TSV format is still supported like before.
* Reorganize the examples folder and add a CSV demo dataset and metadata.
* Implement some integration tests.
* Integrate CodeCov and add the badge to the README.
* Add a Python Quickstart in the README.
* Remake and move the CLI documentation to the sphinx docs folder.
* Move `epochs` arg to `fit` method. See #5.
* Add a `load_demo` function to use in the quickstart.

The new project structure now looks as follows:

```
ctgan
├── __init__.py
├── __main__.py: New CLI module
├── conditional.py: ConditionalGenerator
├── data.py: Functions to write and read data
├── demo.py: Function to load demo data from python
├── models.py: Discriminator, Generator and Residual classes
├── sampler.py: Sampler class
├── synthesizer.py: CTGANSynthesizer class
└── transformer.py: DataTransformer class
examples
├── csv: folder containing a new CSV demo dataset.
└── tsv: folder containing the old `.dat` and `.meta` examples.
tests
└── integration
        └── test_ctgan.py: Simple integration tests
```

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -42,90 +48,144 @@ pip install ctgan
...
+| 38.2477 | Self-emp-inc | 135955    | ... |          40.1124 | Ecuador          | <=50K    |
+| 29.6468 | Private      |   3331.86 | ... |          27.012  | United-States    | <=50K    |
+| 20.9853 | Private      | 120637    | ... |          40.0238 | United-States    | <=50K    |
+|     ... | ...          |       ... | ... |              ... | ...              | ...      |
+
+
+**NOTE**: CTGAN does not distinguish between float and integer columns, which means that it will
+sample float values in all cases. If integer values are required, the outputted float values
+must be rounded to integers in a later step, outside of CTGAN.
+
```
Comment (leix28) [inline README.md]: I think we should provide some evidence to show the performance of CTGAN. For example, the prediction accuracy of CTGAN on demo model.

File: ctgan/synthesizer.py [hunk 2]
```diff
@@ -0,0 +1,209 @@
...
+from torch.nn import functional
+
+from ctgan.conditional import ConditionalGenerator
+from ctgan.models import Discriminator, Generator
+from ctgan.sampler import Sampler
+from ctgan.transformer import DataTransformer
+
+
+class CTGANSynthesizer(object):
+    """docstring for IdentitySynthesizer."""
```
Comment (leix28) [inline ctgan/synthesizer.py]: Add docstring for this class. Clarify the input and output type. For example, both pandas.DataFrame and np.array are supported.

File: README.md (hunk 1 above)
Comment (csala) [reply README.md]: I think it would be better do this on a later PR, to keep avoid making this one bigger.

**Comment by csala** [comment]
> Thank you for refactoring the code. It looks neat now. I add two comments in the code. I assume that saving/restoring the model will happen in next pull request?

Thanks! Yes, the idea is to do that on a separated PR.

Actually, there are a few more changes to do, both in the code and the documentation, like adding the performance or splitting some of the modules in smaller parts, etc. but I prefer doing smaller PRs and releases instead of doing everything at once.

So, if you're OK with it, I will go ahead and merge this one and then open the necessary issues for what's left.

File: ctgan/synthesizer.py (hunk 2 above)
Comment (csala) [reply ctgan/synthesizer.py]: I added docstrings to the class and its public methods, `fit` and `sample`.
More docstrings can be added in a later PR, when the rest of the modules are reviewed.

File: ctgan/synthesizer.py [hunk 3]
```diff
@@ -10,7 +10,28 @@
...
+        gen_dim (tuple or list of ints):
+            Size of the output samples for each one of the Residuals. A Resiudal Layer
+            will be created for each one of the values provided. Defaults to (256, 256).
+        dis_dim (tuple or list of ints):
+            Size of the output samples for each one of the Discriminator Layers. A Linear Layer
+            will be created for each one of the values provided. Defaults to (256, 256).
+        l2scale (float):
+            Wheight Decay for the Adam Optimizer. Defaults to 1e-6.
+        batch_size (int):
+            Number of data samples to process in each epoch.
```
Comment (leix28) [inline ctgan/synthesizer.py]: epoch -> step.
